src/Microsoft.AspNetCore.Mvc.Razor/ViewComponentTagHelpers/ViewComponentTagHelperDescriptorFactory.cs

```diff
@@ -0,0 +1,162 @@
+                tagHelperDescriptors.Add(descriptor);
+            }
+
+            return tagHelperDescriptors;
+        }
+        private TagHelperDescriptor CreateDescriptor(ViewComponentDescriptor viewComponentDescriptor)
+        {
+            var assemblyName = viewComponentDescriptor.TypeInfo.Assembly.GetName().Name;
+            var tagName = GetTagName(viewComponentDescriptor);
+            var typeName = GetTypeName(viewComponentDescriptor);
+
+            var tagHelperDescriptor = new TagHelperDescriptor
+            {
+                TagName = tagName,
+                TypeName = typeName,
+                AssemblyName = assemblyName,
+                TagStructure = TagStructure.NormalOrSelfClosing,
+            };
+
+            SetAttributeDescriptors(viewComponentDescriptor, tagHelperDescriptor);
+            SetRequiredAttributeDescriptors(viewComponentDescriptor, tagHelperDescriptor);
+
+            tagHelperDescriptor.PropertyBag.Add(
+                ViewComponentTagHelperDescriptorConventions.ViewComponentNameKey, viewComponentDescriptor.ShortName);
+            tagHelperDescriptor.PropertyBag.Add(
+                ViewComponentTagHelperDescriptorConventions.ViewComponentTagHelperNameKey,
+                GetTypeName(viewComponentDescriptor));
+
+            return tagHelperDescriptor;
+        }
+
+        private void SetAttributeDescriptors(ViewComponentDescriptor viewComponentDescriptor,
+            TagHelperDescriptor tagHelperDescriptor)
+        {
+            var methodParameters = viewComponentDescriptor.MethodInfo.GetParameters();
+            var attributeDescriptors = new List<TagHelperAttributeDescriptor>();
+
+            foreach (var parameter in methodParameters)
+            {
+                var lowerKebabName = TagHelperDescriptorFactory.ToHtmlCase(parameter.Name);
+                var descriptor = new TagHelperAttributeDescriptor
+                {
+                    Name = lowerKebabName,
+                    PropertyName = parameter.Name,
+                    TypeName = parameter.ParameterType.FullName
+                };
+
+                var tagHelperType = Type.GetType(descriptor.TypeName);
+                if (tagHelperType.Equals(typeof(string)))
+                {
+                    descriptor.IsStringProperty = true;
+                }
+
+                attributeDescriptors.Add(descriptor);
+            }
+
+            tagHelperDescriptor.Attributes = attributeDescriptors;
+        }
+
+        private void SetRequiredAttributeDescriptors(ViewComponentDescriptor viewComponentDescriptor,
+            TagHelperDescriptor tagHelperDescriptor)
+        {
+            var methodParameters = viewComponentDescriptor.MethodInfo.GetParameters();
+            var requiredAttributeDescriptors = new List<TagHelperRequiredAttributeDescriptor>();
+
+            foreach (var parameter in methodParameters)
+            {
+                if (!parameter.HasDefaultValue)
```

Comment:
Once this is done your requiredAttributeDescriptors can be a single liner with a `.Select`